### DISCUSSION TITLE
Pinia index.js is not called when Quasar scaffolding app starts

### DISCUSSION THREAD

**Discussion by annezheng** [quasarframework/quasar#17082] [General - Components / Directives / etc] [answered]
1. I scaffolded quasar V2 project using "yarn create quasar". I chose the following during the setup:

* Quasar CLI, 
* Vue 3,  
* Quasar App CLI with Vite 2, 
* Composition API with <script setup>, 
* pinia,
* axios

2. Inside the created project folder, I ran "quasar info", the entry "vuex - 3.4.0 -- state management for Vue.js" is listed under "Important local packages", see below. 

```
Operating System - Windows_NT(10.0.19045) - win32/x64
NodeJs - 18.20.0

Global packages
  NPM - 10.5.0
  yarn - 1.22.22
  @quasar/cli - 2.1.1
  @quasar/icongenie - Not installed
  cordova - Not installed

Important local packages
  quasar - 2.15.2 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app-vite - 1.8.0 -- Quasar Framework App CLI with Vite
  @quasar/extras - 1.16.11 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.4.21 -- The progressive JavaScript framework for building modern web UI.
  vue-router - 4.3.0
  pinia - 2.1.7 -- Intuitive, type safe and flexible Store for Vue
  vuex - 3.4.0 -- state management for Vue.js
  vite - 2.9.18 -- Native-ESM powered web dev build tool
  eslint - 8.57.0 -- An AST-based pattern checker for JavaScript.
  electron - 9.4.3 -- Build cross platform desktop apps with JavaScript, HTML, and CSS
  electron-packager - 15.1.0 -- Customize and package your Electron app with OS-specific bundles (.app, .exe, etc.) via JS or CLI
  electron-builder - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*

Networking
  Host - azure
  Ethernet - 10.0.0.108
```

2. When I ran "quasar new store demo", the result is this:

```
 App • Generated src/store

 App • Generated store: src\store\demo
 App • Make sure to reference it in src/store/index.js
```

3. I wrote code to access pinia store in boot axios.js. 

* When I ran "quasar dev", pinia stores > index.js is not called by Quasar. I got "getActivePinia() was called but there was no active Pinia. Are you trying to use a store before calling app.use(pinia)?"

4. I copied the same app to my other computer, it runs fine using quasar dev. The difference is that there quasar info shows "vuex - Not installed". 

Question

Why quasar info shows vuex while vuex is not installed? How to remove it?

**Comment by Dtsiantaris** [comment]
Can you provide all the options you inserted into the create quasar

**Comment by annezheng** [comment]
The following are the steps that lead to the issue.

1. Scaffold quasar project

> yarn create quasar

```
√ What would you like to build? » App with Quasar CLI, let's go!
√ Project folder: ... quasar-project
√ Pick Quasar version: » Quasar v2 (Vue 3 | latest and greatest)
√ Pick script type: » Javascript
√ Pick Quasar App CLI variant: » Quasar App CLI with Vite 2 (stable | v1)
√ Package name: ... quasar-project
√ Project product name: (must start with letter if building mobile apps) ... Quasar App
√ Project description: ... A Quasar Project
√ Author: ... 
√ Pick a Vue component style: » Composition API with <script setup>
√ Pick your CSS preprocessor: » None (the others will still be available)
√ Check the features needed for your project: » State Management (Pinia), axios
? Install project dependencies? (recommended)
>   Yes, use yarn
```

2. Add code to access pinia store in boot > axios.js

```
import { boot } from 'quasar/wrappers'
import axios from 'axios'

import { useCounterStore } from '../stores/example-store' // added

// Be careful when using SSR for cross-request state pollution
// due to creating a Singleton instance here;
// If any client changes this (global) instance, it might be a
// good idea to move this instance creation inside of the
// "export default () => {}" function below (which runs individually
// for each client)
const api = axios.create({ baseURL: 'https://api.example.com' })

export default boot(({ app }) => {
  // for use inside Vue files (Options API) through this.$axios and this.$api

  app.config.globalProperties.$axios = axios
  // ^ ^ ^ this will allow you to use this.$axios (for Vue Options API form)
  //       so you won't necessarily have to import axios in each vue file

  app.config.globalProperties.$api = api
  // ^ ^ ^ this will allow you to use this.$api (for Vue Options API form)
  //       so you can easily perform requests against your app's API

  const cnt = useCounterStore() // added
  console.log('boot axios - useCounterStore, cnt', cnt) // added
})

export { api }

```

3. Add console log in stores > index.js

```
import { store } from 'quasar/wrappers'
import { createPinia } from 'pinia'

/*
 * If not building with SSR mode, you can
 * directly export the Store instantiation;
 *
 * The function below can be async too; either use
 * async/await or return a Promise which resolves
 * with the Store instance.
 */

export default store((/* { ssrContext } */) => {
  const pinia = createPinia()

  // You can add Pinia plugins here
  // pinia.use(SomePiniaPlugin)

  console.log('stores - pinia', pinia) // added

  return pinia
})
```

4. Run "quasar dev", got error in browser console.

```
[Quasar] Running SPA. ---- client-entry.js:47:9

[Quasar] boot error: Error: [🍍]: "getActivePinia()" was called but there was no active Pinia. Are you trying to use a store before calling "app.use(pinia)"? ---- client-entry.js:106:15

See https://pinia.vuejs.org/core-concepts/outside-component-usage.html for help.
This will fail in production.
    useStore pinia.mjs:1709
    <anonymous> axios.js:25
    start client-entry.js:90
    <anonymous> client-entry.js:159
```

5. Run "quasar new store demo", got this:

```
 App • Generated src/store

 App • Generated store: src\store\demo
 App • Make sure to reference it in src/store/index.js
```

This happens after I installed `NVM for Windows` so that I can run both quasar V1 and V2 apps in one PC (Windows 10). Previously I have my quasar V1 app (using vuex, vue 2, webpack) in this PC. The quasar V1 app still runs fine after using NVM. But my quasar V2 app (using pinia, vue 3, vite) encountered this issue.

**Reply by Dtsiantaris** [reply]
I'm a little confused because in your original question you said that pinia did not get installed even though you picked it. 
I guess you solved this?

Anyways as to your question about using pinia. Every pinia instance is accessible only after `app.use(pinia)` has been called. In Quasar this happens 'internally'(Maybe not the right terminology). 
You can have a look at [that](https://pinia.vuejs.org/core-concepts/outside-component-usage.html#Single-Page-Applications) for more info about pinia in general. 
I actually created a project using the same exact options and copying your code. I had no issues and the code worked as intended.
Are you sure you are not accessing the store anywhere else? The usual suspects for such cases are route files. ie:
```
import { useCounterStore } from 'src/stores/example-store'

// This will create the  `"getActivePinia()" was called but there was no active Pinia.` error
const store = useCounterStore()

const routes = [
  {
    path: '/',
    component: () => import('layouts/MainLayout.vue'),
    children: [
      { path: '', component: () => import('pages/IndexPage.vue') }
    ]
  },

  // Always leave this as last one,
  // but you can also remove it
  {
    path: '/:catchAll(.*)*',
    component: () => import('pages/ErrorNotFound.vue')
  }
]

export default routes
```

Are you by any chance using SSR? If so, you need to do this:

```
export default boot(({ app, store }) => {
  **// Pass the exposed instance when using SSR**
  const cnt = useCounterStore(store)
  ....
  ```

**Comment by annezheng** [comment]
I didn't use SSR and I didn't change the router files in the scaffolding quasar project. 

I wonder how Quasar Info finds `Important local packages`.  In my case, vuex is NOT installed in my scaffolding quasar project. But the entry `vuex - 3.4.0 -- state management for Vue.js` is listed there. 

If I install vuex in the project, Quasar Info updates the vuex entry to be `vuex - 4.1.0 -- state management for Vue.js` which is correct. Then if I uninstall vuex, Quasar Info shows `vuex - 3.4.0 -- state management for Vue.js` again. It should be `vuex - Not installed`.

**Reply by yusufkandemir** [reply]
Run `yarn why vuex` to see from which package it does come. If it's not coming from there, it might be because you have it installed globally for some reason.

Also, please use the "Write a reply" box and click the "Reply" button instead of using the "Add your answer here..." box and clicking the "Comment" button.

**Reply by Dtsiantaris** [reply]
Considering this:
`I copied the same app to my other computer, it runs fine using quasar dev. The difference is that there quasar info shows "vuex - Not installed".`
And this: 
`Then if I uninstall vuex, Quasar Info shows vuex - 3.4.0 -- state management for Vue.js again`

I think Yusuf is right and vuex is getting installed from somewhere else.

**Reply by annezheng** [reply]
Sorry for the mistake. The command `yarn why vuex` and `yarn list` didn't give me vuex in the output. But `npm ls` does. 

> yarn why vuex
```
yarn why v1.22.22
[1/4] Why do we have the module "vuex"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
error We couldn't find a match!
Done in 0.27s.
```

The entry `UNMET OPTIONAL DEPENDENCY vuex@^4.0.0` is listed under `@quasar/app-vite@1.8.0`. See line 42 below.

Why am I getting "UNMET OPTIONAL DEPENDENCY vuex@^4.0.0" from @quasar/app-vite? How can I remove it? 

> npm ls --unicode --depth
```
1 ├─┬ @quasar/app-vite@1.8.0
2 │ ├── @quasar/render-ssr-error@1.0.3
3 │ ├── @quasar/vite-plugin@1.6.0
4 │ ├── @rollup/pluginutils@4.2.1
5 │ ├── @types/chrome@0.0.208
6 │ ├── @types/compression@1.7.5
7 │ ├── @types/cordova@0.0.34
8 │ ├── @types/express@4.17.21
9 │ ├── @vitejs/plugin-vue@2.3.4
10│ ├── archiver@5.3.2
11│ ├── chokidar@3.6.0
12│ ├── ci-info@3.9.0
13│ ├── compression@1.7.4
14│ ├── cross-spawn@7.0.3
15│ ├── dot-prop@6.0.1
16│ ├── UNMET OPTIONAL DEPENDENCY electron-builder@>= 22
17│ ├── UNMET OPTIONAL DEPENDENCY electron-packager@>= 15
18│ ├── elementtree@0.1.7
19│ ├── esbuild@0.14.51
20│ ├── UNMET OPTIONAL DEPENDENCY eslint@^8.11.0
21│ ├── express@4.19.2
22│ ├── fast-glob@3.2.12
23│ ├── fs-extra@11.2.0
24│ ├── html-minifier@4.0.0
25│ ├── inquirer@8.2.6
26│ ├── isbinaryfile@5.0.2
27│ ├── kolorist@1.8.0
28│ ├── lodash@4.17.21
29│ ├── minimist@1.2.8
30│ ├── open@8.4.2
31│ ├── pinia@2.1.7 deduped
32│ ├── quasar@2.15.2 deduped
33│ ├── register-service-worker@1.7.2
34│ ├── rollup-plugin-visualizer@5.12.0
35│ ├── sass@1.74.1
36│ ├── semver@7.6.0
37│ ├── serialize-javascript@6.0.2
38│ ├── table@6.8.2
39│ ├── vite@2.9.18
40│ ├── vue-router@4.3.0 deduped
41│ ├── vue@3.4.21 deduped
42│ ├── UNMET OPTIONAL DEPENDENCY vuex@^4.0.0
43│ ├── webpack-merge@5.10.0
44│ └── UNMET OPTIONAL DEPENDENCY workbox-build@>= 6
45├── @quasar/extras@1.16.11
46├─┬ autoprefixer@10.4.19
47│ ├── browserslist@4.23.0
48│ ├── caniuse-lite@1.0.30001606
49│ ├── fraction.js@4.3.7
50│ ├── normalize-range@0.1.2
51│ ├── picocolors@1.0.0
52│ ├── postcss-value-parser@4.2.0
53│ └── postcss@8.4.38 deduped
54├─┬ axios@1.6.8
55│ ├── follow-redirects@1.15.6
56│ ├── form-data@4.0.0
57│ └── proxy-from-env@1.1.0
58├─┬ pinia@2.1.7
59│ ├── UNMET OPTIONAL DEPENDENCY @vue/composition-api@^1.4.0
60│ ├── @vue/devtools-api@6.6.1
61│ ├── UNMET OPTIONAL DEPENDENCY typescript@>=4.4.4
62│ ├── vue-demi@0.14.7
63│ └── vue@3.4.21 deduped
64├─┬ postcss@8.4.38
65│ ├── nanoid@3.3.7
66│ ├── picocolors@1.0.0 deduped
67│ └── source-map-js@1.2.0
68├── quasar@2.15.2
69├─┬ vue-router@4.3.0
70│ ├── @vue/devtools-api@6.6.1 deduped
71│ └── vue@3.4.21 deduped
72└─┬ vue@3.4.21
73  ├── @vue/compiler-dom@3.4.21
74  ├── @vue/compiler-sfc@3.4.21
75  ├── @vue/runtime-dom@3.4.21
76  ├── @vue/server-renderer@3.4.21
77  ├── @vue/shared@3.4.21
78  └── UNMET OPTIONAL DEPENDENCY typescript@*
```

**Reply by yusufkandemir** [reply]
It's an optional dependency, meaning, well, it's optional. So, it doesn't have to be met. It means it's not installed. So, you don't have to remove it.

Looking at it here, the problem is likely that it's installed globally. As you don't need it globally, you can just execute the global remove commands to remove it if exists:
```
npm uninstall -g vuex
yarn global remove vuex
```
If it still doesn't work, then I would suggest installing a different Node version using NVM for Windows. I also would recommend using fnm instead of nvm, btw.

**Reply by annezheng** [reply]
1. I executed the global removal commands, the output is below. 

> npm uninstall -g vuex
```
up to date in 494ms
```

> yarn global remove vuex
```
yarn global v1.22.22
[1/2] Removing module vuex...
error This module isn't specified in a package.json file.
info Visit https://yarnpkg.com/en/docs/cli/global for documentation about this command.
```

2. I currently have two node versions installed with nvm:

> nvm list
```
  * 18.20.0 (Currently using 64-bit executable)
    14.21.3
```    

3. Node 14.21.3 is used to run my quasar V1 app and the app runs fine. Note its `npm ls` output has NO vuex entry in "@quasar/app-vite@1.8.0". See below. 

```
1 ├─┬ @quasar/app-vite@1.8.0
2 │ ├── @quasar/render-ssr-error@1.0.3
3 │ ├── @quasar/vite-plugin@1.6.0
4 │ ├── @rollup/pluginutils@4.2.1
5 │ ├── @types/chrome@0.0.208
6 │ ├── @types/compression@1.7.5
7 │ ├── @types/cordova@0.0.34
8 │ ├── @types/express@4.17.21
9 │ ├── @vitejs/plugin-vue@2.3.4
10│ ├── archiver@5.3.2
11│ ├── chokidar@3.6.0
12│ ├── ci-info@3.9.0
13│ ├── compression@1.7.4
14│ ├── cross-spawn@7.0.3
15│ ├── dot-prop@6.0.1
16│ ├── elementtree@0.1.7
17│ ├── esbuild@0.14.51
18│ ├── express@4.19.2
19│ ├── fast-glob@3.2.12
20│ ├── fs-extra@11.2.0
21│ ├── html-minifier@4.0.0
22│ ├── inquirer@8.2.6
23│ ├── isbinaryfile@5.0.2
24│ ├── kolorist@1.8.0
25│ ├── lodash@4.17.21
26│ ├── minimist@1.2.8
27│ ├── open@8.4.2
28│ ├── register-service-worker@1.7.2
29│ ├── rollup-plugin-visualizer@5.12.0
30│ ├── sass@1.74.1
31│ ├── semver@7.6.0
32│ ├── serialize-javascript@6.0.2
33│ ├── table@6.8.2
34│ ├── vite@2.9.18
35│ └── webpack-merge@5.10.0
36├── @quasar/extras@1.16.11
37├─┬ autoprefixer@10.4.19
38│ ├── browserslist@4.23.0
39│ ├── caniuse-lite@1.0.30001606
40│ ├── fraction.js@4.3.7
41│ ├── normalize-range@0.1.2
42│ ├── picocolors@1.0.0
43│ └── postcss-value-parser@4.2.0
44├─┬ axios@1.6.8
45│ ├── follow-redirects@1.15.6
46│ ├── form-data@4.0.0
47│ └── proxy-from-env@1.1.0
48├─┬ pinia@2.1.7
49│ ├── @vue/devtools-api@6.6.1
50│ └── vue-demi@0.14.7
51├─┬ postcss@8.4.38
52│ ├── nanoid@3.3.7
53│ ├── picocolors@1.0.0 deduped
54│ └── source-map-js@1.2.0
55├── quasar@2.15.2
56├─┬ vue@3.4.21
57│ ├── @vue/compiler-dom@3.4.21
58│ ├── @vue/compiler-sfc@3.4.21
59│ ├── @vue/runtime-dom@3.4.21
60│ ├── @vue/server-renderer@3.4.21
61│ └── @vue/shared@3.4.21
62└─┬ vue-router@4.3.0
63  └── @vue/devtools-api@6.6.1 deduped
```
4. Node 18.20.0 is used to run my quasar V2 app where I have this issue. Its `npm ls` output has vuex entry in "@quasar/app-vite@1.8.0". See previous post. 

5. I did comparison of the two, the following entries exist only in `Node 18.20.0 npm ls` output, but not in the "Node 14.21.3 npm ls" output.
```
16 │ ├── UNMET OPTIONAL DEPENDENCY electron-builder@>= 22
17 │ ├── UNMET OPTIONAL DEPENDENCY electron-packager@>= 15

20 │ ├── UNMET OPTIONAL DEPENDENCY eslint@^8.11.0

31 │ ├── pinia@2.1.7 deduped
32 │ ├── quasar@2.15.2 deduped

40 │ ├── vue-router@4.3.0 deduped
41 │ ├── vue@3.4.21 deduped
42 │ ├── UNMET OPTIONAL DEPENDENCY vuex@^4.0.0

44 │ └── UNMET OPTIONAL DEPENDENCY workbox-build@>= 6

53 │ └── postcss@8.4.38 deduped

61 │ ├── UNMET OPTIONAL DEPENDENCY typescript@>=4.4.4

63 │ └── vue@3.4.21 deduped

71 │ └── vue@3.4.21 deduped

78 └── UNMET OPTIONAL DEPENDENCY typescript@*
```

Why "node 14.21.3 npm ls" output has no vuex entry within "@quasar/app-vite@1.8.0" whereas node 18.20.0 output does? 

Which node version do you recommend to install with NVM for windows?

**Reply by Dtsiantaris** [reply]
Did you try a clean install of node 18.20? I would also suggest trying in a way that NVM is not involved. I believe this is what causes your problem essentially.

**Reply by annezheng** [reply]
Can you please elaborate on "clean install node"? 

I had uninstalled NVM and yarn, checked and manually deleted AppData\Local\npm-cache, AppData\Local\Yarn\Cache, then reinstalled NVM, yarn and node 18.20.0. But the issue remains.

**Reply by Dtsiantaris** [reply]
Remove every node version. Delete NVM. Then reinstall only node 18.20 without NVM. If nothing changes remove it again and try a different node version(A higher one, ie 19). 
The problem might be either something to do with NVM or specific node version. You can also try fnm like Yusuf suggested instead of NVM.

**Reply by annezheng** [reply]
I will try it. Thanks for you and Yusuf's kind help!

**Reply by annezheng** [reply, marked as answer]
I uninstalled NVM for Windows, and only installed node 18.20.2. I tested again but the issue remained. 

I applied the work-around provided [here](https://github.com/quasarframework/quasar/issues/14016), and my scaffolding quasar project started. 

```
In quasar.config.js

sourceFiles: {
 store: 'src/stores/index'
}
```
However, I got `this.router is undefined` when I accessed "this.router" in a pinia store. Once again, this second issue doesn't manifest if I run the same code in another computer.

I found the culprit finally. 

* My scaffolding `quasar-project` is in this path
D:\Programming\JavaScript\my-quasar-v2-projects\quasar-project

* I had a node_modules in this path which contains vuex
D:\Programming\JavaScript\node_modules 

* Once I deleted this node_modules, Quasar Info outputs `vuex - Not installed`, and the quasar V2 project runs problem free.

Is this intended behavior that Quasar Info would also search "node_modules" outside the current project folder for the `Important local packages`?

**Reply by yusufkandemir** [reply]
It's not a Quasar-specific behavior, it's how Node.js resolves modules:
> If it is not found there, then it moves to the parent directory, and so on, until the root of the file system is reached.

https://nodejs.org/api/modules.html#loading-from-node_modules-folders

**Reply by Dtsiantaris** [reply]
Good work! Package managers are useful but tricky sometimes. 
I ran into a 'similar' problem, where yarn hoists dependencies and typescript cannot find their modules.
[Here](https://stackoverflow.com/questions/78303885/typescript-not-resolving-nested-node-modules) and [here](https://github.com/quasarframework/quasar/discussions/17094) if you guys have any idea how I could possibly solve this...
